File: sdk/go/public_app_client_test.go
package gestalt_test

import (
	"context"
	"encoding/base64"
	"encoding/json"
	"errors"
	"os"
	"path/filepath"
	"reflect"
	"testing"

	gproto "google.golang.org/protobuf/proto"

	"github.com/valon-technologies/gestalt/sdk/go/publicclient/generated"
	proto "github.com/valon-technologies/gestalt/server/rpc/protov1/v1"
)

type clientCase struct {
	ID            string         `json:"id"`
	Method        string         `json:"method"`
	PublicRequest map[string]any `json:"publicRequest"`
	WireRequest   map[string]any `json:"wireRequest"`
	Response      struct {
		OperationResult *struct {
			Status     int32          `json:"status"`
			Headers    map[string]any `json:"headers"`
			BodyBase64 string         `json:"bodyBase64"`
		} `json:"operationResult"`
		GestaltError *struct {
			Code    int32  `json:"code"`
			Message string `json:"message"`
		} `json:"gestaltError"`
	} `json:"response"`
	Expect struct {
		Result       any `json:"result"`
		GestaltError *struct {
			Code    int32  `json:"code"`
			Message string `json:"message"`
		} `json:"gestaltError"`
		Calls int `json:"calls"`
	} `json:"expect"`
}

type recordingTransport struct {
	calls       int
	err         error
	body        []byte
	lastRequest gproto.Message
}

func (t *recordingTransport) Unary(_ context.Context, method generated.Method, request, response gproto.Message) error {
	t.calls++
	t.lastRequest = request
	if method.Name != "Invoke" {
		return errors.New("unexpected method: " + method.Name)
	}
	if t.err != nil {
		return t.err
	}
	out, ok := response.(*proto.OperationResult)
	if !ok {
		return errors.New("unexpected response type")
	}
	out.Status = 200
	out.Body = t.body
	return nil
}

func loadClientCases(t *testing.T) []clientCase {
	t.Helper()
	data, err := os.ReadFile(filepath.Join("..", "testdata", "public_conformance", "client_cases.json"))
	if err != nil {
		t.Fatalf("read client_cases.json: %v", err)
	}
	var cases []clientCase
	if err := json.Unmarshal(data, &cases); err != nil {
		t.Fatalf("unmarshal client_cases.json: %v", err)
	}
	return cases
}

func wireRequestJSON(t *testing.T, msg gproto.Message) map[string]any {
	t.Helper()
	raw, err := json.Marshal(msg)
	if err != nil {
		t.Fatalf("marshal wire request: %v", err)
	}
	var out map[string]any
	if err := json.Unmarshal(raw, &out); err != nil {
		t.Fatalf("unmarshal wire request: %v", err)
	}
	return out
}

func TestPublicAppClientSharedCases(t *testing.T) {
	for _, tc := range loadClientCases(t) {
		t.Run(tc.ID, func(t *testing.T) {
			rec := &recordingTransport{}
			switch tc.ID {
			case "invoke_success":
				body, err := base64.StdEncoding.DecodeString(tc.Response.OperationResult.BodyBase64)
				if err != nil {
					t.Fatalf("decode body: %v", err)
				}
				rec.body = body
			case "platform_error":
				rec.err = &generated.GestaltError{
					Code:    generated.GestaltErrorCode(tc.Response.GestaltError.Code),
					Message: tc.Response.GestaltError.Message,
				}
			default:
				t.Fatalf("unknown case %q", tc.ID)
			}

			client := generated.NewAppClient(rec)
			request := &generated.AppInvokeRequest{
				App:       tc.PublicRequest["app"].(string),
				Operation: tc.PublicRequest["operation"].(string),
			}
			if params, ok := tc.PublicRequest["params"].(map[string]any); ok {
				request.Params = params
			}

			switch tc.ID {
			case "invoke_success":
				got, err := client.Invoke(context.Background(), request)
				if err != nil {
					t.Fatalf("Invoke: %v", err)
				}
				if rec.lastRequest == nil {
					t.Fatal("transport did not receive a request")
				}
				gotWire := wireRequestJSON(t, rec.lastRequest)
				if !reflect.DeepEqual(gotWire, tc.WireRequest) {
					t.Fatalf("transport request = %#v, want %#v", gotWire, tc.WireRequest)
				}
				if !jsonEqual(got, tc.Expect.Result) {
					t.Fatalf("result = %#v, want %#v", got, tc.Expect.Result)
				}
			case "platform_error":
				_, err := client.Invoke(context.Background(), request)
				if rec.lastRequest == nil {
					t.Fatal("transport did not receive a request")
				}
				gotWire := wireRequestJSON(t, rec.lastRequest)
				if !reflect.DeepEqual(gotWire, tc.WireRequest) {
					t.Fatalf("transport request = %#v, want %#v", gotWire, tc.WireRequest)
				}
				var gerr *generated.GestaltError
				if !errors.As(err, &gerr) {
					t.Fatalf("Invoke error = %v, want *generated.GestaltError", err)
				}
				if int32(gerr.Code) != tc.Expect.GestaltError.Code || gerr.Message != tc.Expect.GestaltError.Message {
					t.Fatalf("GestaltError = %+v, want code=%d message=%q", gerr, tc.Expect.GestaltError.Code, tc.Expect.GestaltError.Message)
				}
			}

			if rec.calls != tc.Expect.Calls {
				t.Fatalf("transport calls = %d, want %d", rec.calls, tc.Expect.Calls)
			}
		})
	}
}

func jsonEqual(left, right any) bool {
	leftBytes, _ := json.Marshal(left)
	rightBytes, _ := json.Marshal(right)
	return string(leftBytes) == string(rightBytes)
}

File: sdk/typescript/tests/public-app-client.test.ts
import { readFileSync } from "node:fs";
import { join } from "node:path";

import type { Message } from "@bufbuild/protobuf";
import { create } from "@bufbuild/protobuf";
import { expect, test } from "bun:test";

import { AppClient } from "../src/client/generated/app_client.ts";
import { PUBLIC_METHODS } from "../src/client/generated/methods.ts";
import type { PublicAppInvokeRequest } from "../src/client/generated/types.ts";
import type { UnaryTransport } from "../src/client/generated/unary_transport.ts";
import {
  OperationResultSchema,
} from "../src/internal/gen/v1/app_pb.ts";
import { GestaltError, GestaltErrorCode } from "../src/rpc_support.ts";

const fixtureRoot = join(import.meta.dir, "..", "..", "testdata", "public_conformance");

type ClientCase = {
  id: string;
  publicRequest: PublicAppInvokeRequest;
  wireRequest: Record<string, unknown>;
  response: {
    operationResult?: { bodyBase64: string };
    gestaltError?: { code: number; message: string };
  };
  expect: {
    result?: unknown;
    gestaltError?: { code: number; message: string };
    calls: number;
  };
};

class RecordingTransport implements UnaryTransport {
  calls = 0;
  err: GestaltError | undefined;
  body = new Uint8Array();
  lastRequest: Message | undefined;

  async unary<Output extends Message>(
    method: (typeof PUBLIC_METHODS)["app"]["invoke"],
    request: Message,
    _inputSchema: unknown,
    _outputSchema: unknown,
  ): Promise<Output> {
    this.calls += 1;
    this.lastRequest = request;
    if (method.method !== "Invoke") {
      throw new GestaltError(GestaltErrorCode.Internal, "unexpected method");
    }
    if (this.err) {
      throw this.err;
    }
    return create(OperationResultSchema, {
      status: 200,
      body: this.body,
    }) as unknown as Output;
  }
}

function projectWire(
  wire: Record<string, unknown>,
  expected: Record<string, unknown>,
): Record<string, unknown> {
  const out: Record<string, unknown> = {};
  for (const key of Object.keys(expected)) {
    out[key] = wire[key as keyof typeof wire];
  }
  return out;
}

function loadCases(): ClientCase[] {
  return JSON.parse(readFileSync(join(fixtureRoot, "client_cases.json"), "utf8"));
}

test("public app client shared cases", async () => {
  for (const clientCase of loadCases()) {
    const transport = new RecordingTransport();
    if (clientCase.id === "invoke_success") {
      const bodyB64 = clientCase.response.operationResult?.bodyBase64;
      if (!bodyB64) throw new Error("missing body");
      transport.body = Uint8Array.from(atob(bodyB64), (c) => c.charCodeAt(0));
    } else if (clientCase.id === "platform_error") {
      const err = clientCase.response.gestaltError;
      if (!err) throw new Error("missing gestalt error");
      transport.err = new GestaltError(err.code, err.message);
    } else {
      throw new Error(`unknown case ${clientCase.id}`);
    }

    const client = new AppClient(transport);
    const request = clientCase.publicRequest;

    if (clientCase.id === "invoke_success") {
      await expect(client.invoke(request)).resolves.toEqual(clientCase.expect.result);
    } else {
      await expect(client.invoke(request)).rejects.toMatchObject({
        code: clientCase.expect.gestaltError?.code,
        message: clientCase.expect.gestaltError?.message,
      });
    }

    if (!transport.lastRequest) {
      throw new Error("transport did not receive a request");
    }
    const gotWire = projectWire(
      JSON.parse(JSON.stringify(transport.lastRequest)) as Record<string, unknown>,
      clientCase.wireRequest,
    );
    expect(gotWire).toEqual(clientCase.wireRequest);

    expect(transport.calls).toBe(clientCase.expect.calls);
  }
});
